### PR TITLE
Add proxy support for database

### PR DESCRIPTION
**Pull request by IDwrong**
As mentioned in [33#](https://github.com/IceDynamix/reliquary-archiver/issues/33#issuecomment-2183976428), I have implemented proxy support for the database. By default, it doesn't use a proxy, and I've added two command-line parameters to enable it.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -42,7 +48,8 @@ fn main() {
 
     debug!(?args);
 
-    let database = Database::new_from_online();
+    let ureq_agent = UreqAgent::new(args.home_proxy, args.home_proxy);
```
Comment (FungusOVO) [inline src/main.rs:51]: i need this pullrequest, but when i test it, i found this should be 
` let ureq_agent = UreqAgent::new(args.home_proxy, args.system_proxy);`

Comment (IDwrong) [reply src/main.rs:51]: Thanks for your review! I'll fix it tonight.

File: src/export/fribbels.rs [hunk 2]
```diff
@@ -244,6 +247,49 @@ impl Exporter for OptimizerExporter {
...
+            let proxy_enable: u32 = internet_settings.get_value("ProxyEnable").unwrap();
+            let proxy_server: String = internet_settings.get_value("ProxyServer").unwrap();
+
+            if proxy_enable == 0 {
+                Agent::new()
+            } else {
+                AgentBuilder::new()
+                    .proxy(Proxy::new(proxy_server).unwrap())
+                    .build()
+            }
```
Comment (IceDynamix) [inline src/export/fribbels.rs:276]: codeblocks are untested, please recheck if the implementation works out.
```suggestion
            // Open registry key.
            let proxy_server = get_proxy_server_from_registry();
            
            if let Some(proxy_server) = proxy_server {
                AgentBuilder::new()
                    .proxy(Proxy::new(proxy_server).unwrap())
                    .build()
            } else {
                Agent::new()
            }
```

```rs
#[cfg(target_os = "windows")]
fn get_proxy_server_from_registry() -> Option<String> {
    // Open registry key.
    let hkcu = RegKey::predef(HKEY_CURRENT_USER);
    let internet_settings = hkcu.open_subkey_with_flags(
            "Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings",
            KEY_READ,
        )?;
    // Read Proxy Settings.
    let proxy_enable: u32 = internet_settings.get_value("ProxyEnable")?;
    if proxy_enable == 0 {
        return None;
    }

    let proxy_server: String = internet_settings.get_value("ProxyServer")?;
    Some(proxy_server)
}

#[cfg(not(target_os = "windows"))]
fn get_proxy_server_from_registry() -> Option<String> {
    warn!("system proxy is only supported on windows! defaulting to no proxy.")
    None
}
```

File: Cargo.toml [hunk 3]
```diff
@@ -20,6 +20,8 @@ serde_json = "1.0.117"
 tracing = "0.1.40"
 tracing-subscriber = { version = "0.3.18", features = ["env-filter", "json"] }
 ureq = { version = "2.9.7", features = ["json"] }
+winreg = "0.8"
```
Comment (IceDynamix) [inline Cargo.toml:23]:
```suggestion

[target.'cfg(windows)'.dependencies]
winreg = "0.8"
```

File: src/main.rs [hunk 4]
```diff
@@ -32,6 +32,12 @@ struct Args {
     /// Path to output log to
     #[arg(short, long)]
     log_path: Option<PathBuf>,
+    /// Use HOME_PROXY environment variable.
+    #[arg(long, action = clap::ArgAction::SetTrue)]
+    home_proxy: bool,
+    /// Use system proxy. Ignored if home_proxy is set.
```
Comment (IceDynamix) [inline src/main.rs:38]:
```suggestion
    /// Use system proxy. Ignored if home_proxy is set. Only works on Windows.
```

**Review by IceDynamix — CHANGES_REQUESTED** [review]
this fails to build on linux from `winreg` dependency. we do surprisingly have linux users and they don't deserve a failing build :)

the big codeblocks are untested (im just writing this from github)

File: src/export/fribbels.rs [hunk 5]
```diff
@@ -26,6 +26,9 @@ use reliquary::resource::text_map::TextMap;
 use serde::{Deserialize, Serialize};
 use serde::de::DeserializeOwned;
 use tracing::{debug, info, info_span, instrument, trace, warn};
+use ureq::{Agent, AgentBuilder, Proxy};
+use winreg::enums::*;
+use winreg::RegKey;
```
Comment (IceDynamix) [inline src/export/fribbels.rs:31]:
```suggestion
#[cfg(target_os = "windows")]
use winreg::enums::*;
#[cfg(target_os = "windows")]
use winreg::RegKey;
```

**Comment by IDwrong** [comment]
In the last two commits, I was trying to verify them. Sorry if this causes any inconvenience.

**Comment by IceDynamix** [comment]
superceded by #56

**Pull request closed without merging**